### PR TITLE
Add EmailJobErrorHandler to support sending email when job error

### LINKED ISSUE #1440: Add EmailJobErrorHandler to support sending email when job error.
To support for sending the emails when the job executes with errors, we decide to do some extensions for JobErrorHandler.
- create a new module - elasticjob-error-handler/elasticjob-error-handler-email
- load config from System parameters.
- load config from specified conf.

### PR DESCRIPTION
**Pull request by luky116**
Fixes #1440 .

Changes proposed in this pull request:
- add email job handler
-
-

### CODE REVIEW

**Comment by terrymanu** [comment]
Can you change the title?

File: elasticjob-error-handler/elasticjob-error-handler-email/pom.xml [hunk 1]
```diff
@@ -27,4 +27,38 @@
...
+
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-inline</artifactId>
+        </dependency>
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/pom.xml]: "Test" scope is required.

File: elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/ConfigurationLoader.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import lombok.NoArgsConstructor;
+import org.apache.shardingsphere.elasticjob.infra.yaml.YamlEngine;
+
+import java.io.InputStream;
+
+/**
+ * Job error configuration loader.
+ */
+@NoArgsConstructor
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/ConfigurationLoader.java]: "Private" access required.

File: elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    @Override
+    public void handleException(final String jobName, final Throwable cause) {
+        try {
+            Preconditions.checkNotNull(emailConfiguration);
+            String content = buildContent(jobName, cause);
+            Message message = buildMessage(content);
+            sendMessage(message);
+        } catch (final NullPointerException | MessagingException ex) {
+            log.error("Elastic job: email job handler error", ex);
+        }
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandler.java:67]: Catch 'NullPointerException' seems weird.

File: elasticjob-error-handler/elasticjob-error-handler-email/src/main/resources/META-INF/services/org.apache.shardingsphere.elasticjob.error.handler.JobErrorHandler [hunk 4]
```diff
@@ -15,6 +15,4 @@
 # limitations under the License.
 #
 
-org.apache.shardingsphere.elasticjob.infra.handler.error.impl.LogJobErrorHandler
-org.apache.shardingsphere.elasticjob.infra.handler.error.impl.IgnoreJobErrorHandler
-org.apache.shardingsphere.elasticjob.infra.handler.error.impl.ThrowJobErrorHandler
+org.apache.shardingsphere.elasticjob.error.handler.email.EmailJobErrorHandler
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/src/main/resources/META-INF/services/org.apache.shardingsphere.elasticjob.error.handler.JobErrorHandler]: Need a blank line.

File: elasticjob-error-handler/elasticjob-error-handler-general/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/JobErrorHandlerFactory.java [hunk 5]
```diff
@@ -56,5 +56,5 @@ public static JobErrorHandler getHandler(final String type) {
             throw new JobConfigurationException("Can not find job error handler type '%s'.", type);
         }
         return HANDLERS.get(type);
-    } 
+    }
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-general/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/JobErrorHandlerFactory.java]: Unnecessary changes.

File: elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/JobErrorHandlerFactoryTest.java [hunk 6]
```diff
@@ -7,7 +7,7 @@
  * the License.  You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
- *  
+ *
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/JobErrorHandlerFactoryTest.java]: Unnecessary changes.

File: elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/IgnoreJobErrorHandlerTest.java [hunk 7]
```diff
@@ -7,7 +7,7 @@
  * the License.  You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
- *  
+ *
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/IgnoreJobErrorHandlerTest.java]: Unnecessary changes.

File: elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/IgnoreJobErrorHandlerTest.java [hunk 8]
```diff
@@ -20,7 +20,7 @@
 import org.junit.Test;
 
 public final class IgnoreJobErrorHandlerTest {
-    
+
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-general/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/IgnoreJobErrorHandlerTest.java]: Unnecessary changes.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

**Comment by luky116** [comment]
> Can you change the title?

I'll change it

File: elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandler.java (hunk 3 above)
Comment (luky116) [reply elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandler.java:67]: if user don't set email configuration, `Preconditions.checkNotNull(emailConfiguration)` will throws NullPointerException,it menas user don't want email handler. So here will catch NullPointerException.

Comment (vran-dev) [reply elasticjob-error-handler/elasticjob-error-handler-email/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandler.java:67]: hello, I hava some suggestions you can refre to.

if `emailConfiguration` is null means do nothing, i think use `check and  (log?) return`  is better than `check and catch`,  because
 1.  code taste: manually check and catch could increased code complexity
 2.  performance:  avoid use exception as control flow

File: elasticjob-error-handler/elasticjob-error-handler-email/src/test/resources/error-handler-email.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+email:
+  host: test.mail.com
+  port: 123
+  username: username
+  password: password
+  protocol: smtp
+  from: testmail@qiyi.com
+  to: xxx1@ejob.com
+  cc: xxx2@ejob.com
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/src/test/resources/error-handler-email.yaml:26]: Need a blank line.

File: elasticjob-error-handler/elasticjob-error-handler-email/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    }
+        
+    @Test
+    @SneakyThrows
+    public void assertHandleExceptionForNullConfiguration() {
+        EmailJobErrorHandler emailJobErrorHandler = new EmailJobErrorHandler();
+        Field emailConfigurationField = EmailJobErrorHandler.class.getDeclaredField("emailConfiguration");
+        emailConfigurationField.setAccessible(true);
+        emailConfigurationField.set(emailJobErrorHandler, null);
+
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-email/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/email/EmailJobErrorHandlerTest.java:52]: Remove redundant blank line.

**Review by TeslaCN — CHANGES_REQUESTED** [review]
